[FX-572] Add customer-perceived response metrics

djoksimo opened the pull request:
<!-- Update your title to prefix with your ticket number -->

## What

<!-- Describe your changes here -->

- Add customer-perceived response metrics to `checkBalance` and `capturePayment`

<!-- If you are making a front-end change, please include a screen recording and post it in #feature-recordings -->

## Why

https://linear.app/joinforage/issue/FX-572/ios-include-customer-perceived-metric-in-monitoring

<!-- Describe the motivations behind this change if they are a subset of your ticket -->

## Test Plan

<!-- IMPORTANT: QA Tests and Unit Tests must be passed locally before this PR can be merged. -->

- ✅ | ❌ Not a front-end change
- ✅ | ❌ iOS QA Tests passed locally
- ✅ | ❌ Unit Tests passed locally

## How

<!-- Describe the rollout plan if it includes multiple PRs/Repos or requires extra steps beyond rolling back the Service -->

- PATCH-level change
- Will follow-up with a cleanup PR for `checkBalance` and `capturePayment` ([FX-631](https://linear.app/joinforage/issue/FX-631/ios-clean-up-checkbalance-and-capturepayment))

djoksimo commented:
Current dependencies on/for this PR:
* main
  * **PR #118** <a href="https://app.graphite.dev/github/pr/teamforage/forage-ios-sdk/118" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/teamforage/forage-ios-sdk/118?utm_source=stack-comment).

dleis612 commented on Sources/ForageSDK/Foundation/ForageSDK+ForageServices.swift:
```diff
@@ -94,7 +94,16 @@ extension ForageSDK: ForageSDKService {
+            // This block is used for tracking important Metrics!
+            // -----------------------------------------------------
+            let responseMonitor = CustomerPerceivedResponseMonitor.newMeasurement(
+                vaultType: foragePinTextField.collector!.getVaultType(),
```
Any way to avoid this unwrap or are we just gonna roll the dice?

djoksimo replied:
We used to force unwrap `collector` in this other part of the same method – so the force-wrapping from my diff does not make this method any more prone to crashes than it already was.

However, i think we can refactor the force unwrapping and handle this more gracefully; I'll handle this as part of this cleanup ticket that [is on my plate for this week](https://linear.app/joinforage/issue/FX-631/ios-clean-up-checkbalance-and-capturepayment)

djoksimo commented:
I've tested my changes in Datadog against dev & staging

- Seems [like we use](https://github.com/teamforage/forage-ios-sdk/blob/46aad2309319f5d03bfe3796782942b032c5933a/Sources/ForageSDK/Foundation/Network/Provider.swift#L157) `code: "error"` for JSON decoding errors. If we changed this code, this could be a breaking change, so I've added code: "error" to our list of things that we treat as "failures"

Here are some DD links that demonstrate the correctness of these logs
- [Logs](https://app.datadoghq.com/logs?query=service%3Aios-sdk%20%40log_type%3Ametric%20&cols=host%2Cservice&index=%2A&messageDisplay=inline&refresh_mode=paused&stream_sort=desc&viz=stream&from_ts=1697494917368&to_ts=1697495064592&live=false)
- [Metrics explorer](https://app.datadoghq.com/metric/explorer?start=1697487650000&end=1697489471571&paused=true#N4Ig7glgJg5gpgFxALlAGwIYE8D2BXJVEADxQEYAaELcqyKBAC1pEbghkcLIF8qo4AMwgA7CAgg4RKUAiwAHOChASAtnADOcAE4RNIKtrgBHPJoQaUAbVBGN8qVoD6gnNtUZCKiOq279VKY6epbINiAiGOrKQdpYZAYgUJ4YThr42gDGSsgg6gi6mZaBZnHKGniqyBpQANYAdJl4Ggg46tpOilnsAG5wUPV2DiJajfgiCMBwIj3IAj0ABBQLfj0Q2ciSGgC0NbU8CwBGWAtTfRNO+AiZbXAUrtoY8E462m5ONwIUPRh4aAhOOSKCgYTISKQ8eoYDQfcYIAAUAEoQDwALpUB6qP7Q6ygTHYmKleIo9EgFpYNA5UDyDCUhAIHJJHAwD6YDQadaJNCiOCAhSM9LcqBcnlOehMZQiNweNAo-gQeyYLB8xTKQ4YbSWNE8Phk+TchAAYSkwhgKBEfzQPCAA)
- [Error codes](https://app.datadoghq.com/logs?query=service%3Aios-sdk%20%40log_type%3Ametric%20%40event_name%3Acustomer_perceived_response%20&agg_q=%40forage_error_code&cols=host%2Cservice&index=%2A&messageDisplay=inline&refresh_mode=paused&sort_m=&sort_t=&stream_sort=desc&top_n=10&top_o=top&viz=toplist&x_missing=true&from_ts=1697470204920&to_ts=1697495236000&live=false)
- [Vault Dashboard](https://app.datadoghq.com/dashboard/82m-czd-39c/vaulting-metrics?refresh_mode=paused&tpl_var_env%5B0%5D=dev&tpl_var_env%5B1%5D=sandbox&tpl_var_env%5B2%5D=staging&from_ts=1697466867553&to_ts=1697495346000&live=false)
- [Customer Dashboard](https://app.datadoghq.com/dashboard/ivu-m4y-3dy/customer-perception-metrics?refresh_mode=paused&tpl_var_env%5B0%5D=dev&tpl_var_env%5B1%5D=staging&tpl_var_env%5B2%5D=sandbox&tpl_var_sdk%5B0%5D=ios-sdk&from_ts=1697480489815&to_ts=1697495305311&live=false)

dleis612 commented:
> * like we use

Yuck, both this line and the `invalid_data` codes are terrible leftover tech debt from me + mobile devs. We should really consider getting rid of them ASAP.

dleis612 commented on Sources/ForageSDK/Foundation/Telemetry/CustomerPerceivedResponseMonitor.swift:
```diff
@@ -0,0 +1,84 @@
+    
+    // override to set event_outcome to "failure" if we know the event has a forage_error_code
+    @discardableResult
+    internal override func setForageErrorCode(_ error: Error) -> ResponseMonitor {
```
I think you may want to return a `CustomerPerceivedResponseMonitor` object here.

djoksimo replied:
ty for catching this 8ca2ef9666f652ab91ec52512761d9c125c9ff08

miketikh reviewed (approved):
Looks great! Like the extensibility of building the Customer Perceived and Vault monitors on top of ResponseMonitor
